feat: Add logs to helm test

jeff-knurek opened the pull request:
**What this PR does / why we need it**:
There's a Helm 3 proposal for improvements to `helm test` #6020 but in the meantime there is still a lot of demand for being able to view the pod logs from test, while still taking advantage of the `--cleanup` flag. 

**Special notes for your reviewer**:
I saw that this PR was opened and closed: https://github.com/helm/helm/pull/4081
But only after finishing the work did I find that this one is still open: https://github.com/helm/helm/pull/3907

This particular solution is to only attempt to dump the full logs at the end of the tests, and does not make an attempt at streaming logs. So this solution is not impacted by parallel tests.

**If applicable**:
- [x] this PR contains documentation
- [x] this PR contains unit tests
- [x] this PR has been tested for backwards compatibility

closes #2923
refs #1957

Commit:
TEST: add option to dump logs during helm test

Signed-off-by: Jeff Knurek <j.knurek@travelaudience.com>

Commit:
TEST: cleanup code

Signed-off-by: Jeff Knurek <j.knurek@travelaudience.com>

Commit:
TEST: add some rudimentary test code for helm test logs

Signed-off-by: Jeff Knurek <j.knurek@travelaudience.com>

Commit:
DOC: update documentation for helm test --logs

Signed-off-by: Jeff Knurek <j.knurek@travelaudience.com>

jeff-knurek commented:
**To help with PR review**


The current functionality: (with `--cleanup`)
* `helm install stable/wordpress`
* `helm test billowing-lamb --cleanup`
```
RUNNING: billowing-lamb-credentials-test
FAILED: billowing-lamb-credentials-test, run `kubectl logs billowing-lamb-credentials-test --namespace default` for more info
RUNNING: billowing-lamb-mariadb-test-jkg8v
FAILED: billowing-lamb-mariadb-test-jkg8v, run `kubectl logs billowing-lamb-mariadb-test-jkg8v --namespace default` for more info
Error: 2 test(s) failed
```
* `kubectl logs billowing-lamb-mariadb-test-jkg8v --namespace default`
```
Error from server (NotFound): pods "billowing-lamb-mariadb-test-jkg8v" not found
```

----
New functionality: (with `--cleanup` & `--logs`)
* `helm install stable/wordpress`
* `HELM_HOST=localhost:44134 helm test womping-mole --cleanup --logs`
```
RUNNING: womping-mole-mariadb-test-it5vd
FAILED: womping-mole-mariadb-test-it5vd
RUNNING: womping-mole-credentials-test
FAILED: womping-mole-credentials-test

POD LOGS: womping-mole-mariadb-test-it5vd
1..1
not ok 1 Testing MariaDB is accessible
# (in test file /tests/run.sh, line 2)
#   `mysql -h womping-mole-mariadb -uroot -p$MARIADB_ROOT_PASSWORD -e 'show databases;'' failed
# ERROR 2002 (HY000): Can't connect to MySQL server on 'womping-mole-mariadb' (115)


POD LOGS: womping-mole-credentials-test
ERROR 2002 (HY000): Can't connect to MySQL server on 'womping-mole-mariadb' (115)

Error: 2 test(s) failed
```

jeff-knurek commented on pkg/releasetesting/environment.go:
```diff
@@ -96,7 +96,7 @@ func (env *Environment) streamError(info string) error {
 }
 
 func (env *Environment) streamFailed(name string) error {
-	msg := fmt.Sprintf("FAILED: %s, run `kubectl logs %s --namespace %s` for more info", name, name, env.Namespace)
```
now that I've run the comparison of current behavior vs this change, I wonder if this line should remain as is? 
* it's still helpful when `--cleanup` is not used
* people may be expecting it as is, and using it this way
* it might not be so useful, but it doesn't impact anything to leave it as is

???

bacongobbler replied:
I'd keep it as-is if it's irrelevant to the changes involved in this PR.

bacongobbler commented on pkg/kube/client.go:
```diff
@@ -929,6 +929,24 @@ func (c *Client) watchPodUntilComplete(timeout time.Duration, info *resource.Inf
 
+// GetPodLogs takes pod name and namespace and returns the current logs (streaming is NOT enabled).
+func (c *Client) GetPodLogs(name, ns string) (string, error) {
+	client, _ := c.KubernetesClientSet()
```
we should probably check and return if `err` is non-nil here

bacongobbler commented on pkg/kube/client.go:
```diff
@@ -929,6 +929,24 @@ func (c *Client) watchPodUntilComplete(timeout time.Duration, info *resource.Inf
 }
 
+// GetPodLogs takes pod name and namespace and returns the current logs (streaming is NOT enabled).
+func (c *Client) GetPodLogs(name, ns string) (string, error) {
```
Any way we can return an `io.ReadCloser` instead of a `string` here and let `GetLogs` handle the stream? That way, the consumer (`GetLogs`) can determine what to do with the buffer instead of always reading it into memory. For example, what if the consumer just wanted to open the stream to test that it was possible (like in a test case)?

bacongobbler commented:
Hey @jeff-knurek, just letting you know that as per [the announcement on the mailing list](https://lists.cncf.io/g/cncf-helm/topic/last_v2_feature_release/34301238), Helm 2.15.0 is the last release where we will be merging new features for Helm 2. If you want to get this in for the next release, please let us know and we can work with you to get this in. If not, we'll have to close this and continue this effort over in Helm 3.

Thanks!

jeff-knurek commented:
yeah, would be great to get it into `2.15` - I'll have a go at the two change requests

Commit:
return error is client not set

Signed-off-by: Jeff Knurek <j.knurek@travelaudience.com>

Commit:
change GetPodLogs to return io.Reader instead of string

well, more specifically returns an io.ReadCloser (giving the consumer more capabilities)

Signed-off-by: Jeff Knurek <j.knurek@travelaudience.com>

jeff-knurek commented:
hmm, tests pass for me locally, and there's not enough info to see why they fail here
```
.circleci/test.sh
Running 'make build'
GOBIN=/go/src/k8s.io/helm/bin go install  -tags '' -ldflags '-w -s -X k8s.io/helm/pkg/version.GitCommit=4196766ff2976e0b735e6f520e95ff1725925f20 -X k8s.io/helm/pkg/version.GitTreeState=clean' k8s.io/helm/cmd/...
Running 'unit tests'

==> Running unit tests <==
HELM_HOME=/no/such/dir go test  -run . ./pkg/... ./cmd/... 
....
ok  	k8s.io/helm/pkg/helm/helmpath	0.095s
/usr/local/go/pkg/tool/linux_amd64/link: signal: killed
ok  	k8s.io/helm/pkg/helm/portforwarder	0.036s
....
ok  	k8s.io/helm/pkg/tiller	0.064s
FAIL	k8s.io/helm/pkg/tiller/environment [build failed]
ok  	k8s.io/helm/pkg/timeconv	0.271s
...

Makefile:115: recipe for target 'test-unit' failed
make: *** [test-unit] Error 1
Exited with code 2
```

bacongobbler reviewed (commented):
Looks like there's [one more comment to address](https://github.com/helm/helm/pull/6167#discussion_r329748511) before I approve this. Overall this looks good to me!

bacongobbler commented:
ping

bacongobbler commented:
Hi @jeff-knurek. Me again. :)

Given the low volume of PRs left on the queue, we are planning to cut a 2.15 within the next few business days. Any chance you can resolve the merge conflicts and address previous comments by EOD tomorrow? If not, unfortunately we will have to close this out and carry on the work over in Helm 3. Thanks!

> hmm, tests pass for me locally, and there's not enough info to see why they fail here

We've been seeing a number of Helm 2 test failures on CircleCI due to resource limitations. We haven't identified the issue, though it doesn't appear to be present in the Helm 3 branch.

bacongobbler commented:
closing in favour of #6612. Thanks!

jeff-knurek commented:
sorry, was on holiday. thanks for cleaning this up and getting it merged in #6612 👍
